### DISCUSSION TITLE
2.0.0

### DISCUSSION THREAD

**Discussion by AntonovAnton** [AntonovAnton/math.evaluation#54] [Announcements]
Added support for compiling mathematical expression strings into executable code, producing a delegate that represents the expression.

Example:
```
var fn = "ln(1/x1 + √(1/(x2*x2) + 1))"
    .Compile(new { x1 = 0.0, x2 = 0.0 }, new ScientificMathContext());

var value = fn(new { x1 = -0.5, x2 = 0.5 });
```

This enhancement allows for efficient evaluation of complex mathematical expressions by converting them into callable delegates.

<hr /><em>This discussion was created from the release <a href='https://github.com/AntonovAnton/math.evaluation/releases/tag/2.0.0'>2.0.0</a>.</em>

**Comment by ajitid** [comment]
Hello! A newbie question from me: Are there any other ways beside using `new { x1 = 0.0, x2 = 0.0 }` in `.Compile()` method? I am essentially calling the snippet you have given above in a pseudocode like this:

```text
var expression = '0'
var fn

function callEveryFrame(newExpression, params) {
	// only compile the expression if it is out of date
	if (expression != newExpression) {
		expression = newExpression
		fn = expression.Compile(params, new MathContext());
	}

 	var value = fn(params);
	return value;
}
```

As you can see in my case that `params` in one frame can be `new { a = 0, b = 0 }` and in the next frame can be `new { a = 0, b = 0, c = 0}`. The number of parameters and their names may change over time. I couldn't pass a dictionary to `.Compile()` as it throws an error. Is there any other way to do this?

**Reply by AntonovAnton** [reply]
Hello, try to use Evaluate method instead of Compilation, it is faster than compilation in your case compilation expression every frame decreasing performance, compilation takes time, it is beneficial to compile the expression beforehand if you plan to evaluate it multiple times, especially for 200 or more iterations. Also evaluation not compilation allows you to change the list of params because you don't need compiling input class, evaluation parsing input every time not once as compilation does.

**Reply by ajitid** [reply]
Thanks @AntonovAnton. I actually posed my question wrong. In my usecase, the compilation of expression due change in expression string would happen very rarely. My use case is exactly this, that I need to call the same compiled expression many times with params (easily >200, I may even call same compiled expression during the lifetime of my program).

Given this, how would I solve the problem I mentioned earlier in my post above?

**Reply by ajitid** [reply]
I also noticed that my pseduocode is wrong. I have just updated it so you'd able to understand it now.

**Reply by AntonovAnton** [reply]
I think only Evaluation, because in C# dinamic params it's an array, their has indexes but not names, you need provide names to parse expression, and this expression should include these variables, it means the expression is changing dinamiclly so I don't think it is possible to compile once. If it different every time anyway you need recompile inputs for Dot Net. Evaluate method works as interpretator parsing at runtime that exactly what you need. Trying to change compiled class causes recompilation. In Mathevaluator Evaluate method is fast it's about 500 ns per operation. about 2000 operation per ms, 2000000 per second

**Reply by AntonovAnton** [reply]
Oh, I see you mentioned dictionary, sorry Mathevaluator doesn't support dictionary as an input parameter at the moment, only objects

**Reply by AntonovAnton** [reply]
I think it's possible to add, could you please provide examples of expression and a dictionary that you could pass as input and what you expect to get as a result. I will add it to backlog for future improvements

**Reply by ajitid** [reply]
Thank you. I want to use MathEvaluator to do calculations using DotNetStandartMathContext. Here's an example:

```cs
var dict = new Dictionary<string, double>();
dict.Add("x1", 3.0);
dict.Add("x2", 2.0);

var fn = "x1 + Math.Sin(x2) * 0.5"
    .Compile(dict, new DotNetStandartMathContext());

var value = fn(dict);
Console.WriteLine(value); // 3.4546487...

dict.Add("x1", 6.0);
dict.Add("x2", 4.0);
value = fn(dict);
Console.WriteLine(value); // 5.6215987...

// new expression:

dict.Clear();
dict.Add("x1", 3.0);
dict.Add("x3", 2.0);
dict.Add("x4", 4.0);

fn = "x1 + Math.Sin(x3) * Math.Cos(x4)"
    .Compile(dict, new DotNetStandartMathContext());

value = fn(dict);
Console.WriteLine(value); // 2.4056435...
```

PS: By the way, is Standard intentionally misspelt in DotNetStandartMathContext?

**Reply by AntonovAnton** [reply]
Oh, it's my mistyping I think :) I would wonder next few weeks I don't have enough time to work on the library, I can do it only in free time, I will try my best, sorry

**Reply by AntonovAnton** [reply]
Thanks for example it helps me a lot to improve the library

**Reply by AntonovAnton** [reply]
Hello @ajitid! Added in version [2.3.0](https://github.com/AntonovAnton/math.evaluation/releases/tag/2.3.0)

**Reply by ajitid** [reply]
Thank you!
